JabRef/jabref
# Directory pattern bracketed expressions

- [ x] Change in CHANGELOG.md described
- [ x] Tests created for changes
- [ ] Screenshots added (for bigger UI changes)
- [ x] Manually tested changed features in running JabRef
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)
- [ ] If you changed the localization: Did you run `gradle localizationUpdate`?

The offered PR implements functionality discussed in my "File linking functionality update?" proposal (http://discourse.jabref.org/t/file-linking-functionality-update/723). It is now possible to use bracketed expressions (like "[bibtexkey]" or "[year]\_[auth]\_[firstpage]") consistently in:

- BibTex key generator patterns;
- File search regular expressions (for automatic file linking);
- Import file name patterns;
- Import directory patterns.

To maintain consistent formatting, I have created a new BracketedExpressionExpander class which provides its main methods, 'expandBrackets()'. They essentially contain all bracket expansion functionality moved from BibtexKeyPatternUtil, and the BibtexKeyPatternUtil class now extends BracketedExpressionExpander, and uses its static methods. The BracketedExpressionExpander is also used in FileUtil and RegExpBasedFileFinder, avoiding the previous duplication of functionality in BibtexKeyPatternUtil.makeLabel(...) and RegExpBasedFileFinder.getFieldAndFormat(...) methods (both now use BracketedExpressionExpander.expandBrackets() for dealing with *all* aspects of bracket expansion in one place).

The bracketed expressions now can be used for Import File name patterns and Import File Directory patterns, providing consistent behavior of all "pattern" tabs in the JabRef "Preferences" dialogue; it is also possible to use underscores in the Import File and Directory patterns (e.g. "[year]\_[auth]\_[pages]") and to place Import destination directories deeper in the tree (e.g. "by-author/[authIni1]/[auth]" Directory Pattern now works as expected, creating "by-author/P/Pezzotti/2013\_Pezzotti\_211301.pdf" for the reference: Pezzotti, G. Raman spectroscopy of piezoelectrics Journal of Applied Physics, 2013, 113, 211301).

To retain functionality that was previously implemented by the '\entrytype' Layout pattern, a new bracketed expression "[entrytype]" was created and tested. It is suggested that Layout manager is used exclusively for describing layouts in citation lists, and bracketed expressions are used for describing file system and database entry patterns.

The functionality was tested under LinuxMint 18.1 and Ubuntu 16.04. Unfortunately, I could not test it on Windows hosts since I do not have any of them available for Java testing. I have put in provisions to deal with both '/' and '\' directory separators in Directory Patterns, but I would be grateful if someone checks them on a Windows system.

Sincerely,
Saulius
